README.md
# vagrant-tools
A collection of Python scripts to manage Vagrant boxes in a private repository

## DOCUMENTATION

### Private repository configuration

The private repository can be configured with a [NGINX][1] server, ssh and a
unprivileged user. Basically the server is configured to serve the metadata.json
 file located in the box folder so Vagrant can get box information like versions
 and URLs for them.

As shown below the boxes are stored in /usr/share/nginx/html/vagrant with the
following details:

  - __Box path:__ /usr/share/nginx/html/vagrant/boxname
  - __Metadata path:__ /usr/share/nginx/html/vagrant/boxname/metadata.json
  - __Boxes files:__ /usr/share/nginx/html/vagrant/boxname/boxes/*

__Example of metadata.json__

```json
{
  "description": "This is a example.",
  "name": "amontalban/superbox",
  "versions": [
    {
      "providers": [
        {
          "checksum": "bf74bdbabada3751a91399ab3069d02ed7fa5d0b",
          "checksum_type": "sha1",
          "name": "virtualbox",
          "url": "http://devops.domain.com/vagrant/amontalban/superbox-1.0/boxes/superbox-1.0.box"
        }
      ],
      "version": "1.0"
    }
  ]
}
```

__NGINX server configuration template__
```
server {
    listen 80 default_server;
    listen [::]:80 default_server ipv6only=on;

    root /usr/share/nginx/html;
    index metadata.json;

    server_name devops.domain.com;

    # Match the box name in location and search for its catalog
    # e.g. http://www.example.com/vagrant/devops/ resolves to
    # /var/www/vagrant/devops/devops.json
    #
    location ~ ^/vagrant/([^\/]+)/$ {
        index metadata.json;
        try_files $uri $uri/ metadata.json =404;
        autoindex off;
    }

    location ~ ^/vagrant/([^\/]+)/boxes/$ {
        try_files $uri $uri/ =404;
        autoindex off;
    }

    # Serve json files with content type header application/json
    location ~ \.json$ {
        add_header Content-Type application/json;
    }

    # Serve box files with content type application/octet-stream
    location ~ \.box$ {
        add_header Content-Type application/octet-stream;
    }

    # Deny access to document root and the vagrant folder
    location ~ ^/(vagrant/)?$ {
        return 403;
    }

}
```

The other thing you will need is a user with SSH access to the server and write
rights in the configured path (/usr/share/nginx/html in this case). It's
recommended to use public keys for SSH so you don't have to use passwords. For
more info check this [guide][2].
### vagrant-upload

For more information about this script please check the [documentation][3].

### vagrant-download

For more information about this script please check the [documentation][4].

## CHANGELOG

Check the [CHANGELOG][5] to know about recent changes.

## RELATED LINKS

- [Vagrant box format][6]
- [vube/vagrant-boxer][7]
- [Custom Vagrant Cloud Versioned Box Host][8]

[1]: http://nginx.org
[2]: https://www.digitalocean.com/community/tutorials/how-to-set-up-ssh-keys--2
[3]: https://github.com/amontalban/vagrant-tools/blob/master/VAGRANT-UPLOAD.md
[4]: https://github.com/amontalban/vagrant-tools/blob/master/VAGRANT-DOWNLOAD.md
[5]: https://github.com/amontalban/vagrant-tools/blob/master/CHANGELOG.md
[6]: http://docs.vagrantup.com/v2/boxes/format.html
[7]: https://github.com/vube/vagrant-boxer
[8]: http://blog.el-chavez.me/2015/01/31/custom-vagrant-cloud-host/


CHANGELOG.md
# CHANGELOG

## 0.0.2

__NEW FEATURES:__

   * Created new script called `vagrant-downloader` to download boxes from\
   repository so for example you can use base boxes for new Packer generated
   boxes.

__IMPROVEMENTS:__

   * Update documentation.
   * Add a requirements.txt file to install Python packages.

---

## 0.0.1

__IMPROVEMENTS:__

  * Validate config values
  * Set a message for every step in the process

__BUG FIXES:__

  * SHA1 wasn't calculated properly
  * Configure metadata to not upload complete box file path, just box file name.


vagrant-download.py
#!/usr/bin/env python

import os, sys, argparse, time, tempfile, re, json, shutil, urlparse
import hashlib, tarfile
from urllib2 import Request, urlopen, URLError, HTTPError
from validators import url

#----------------------------------------------------------------------
#-- Script settings

# Default settings
default_provider = "virtualbox"
vm_providers = ['vmware_workstation', 'vmware_fusion', 'virtualbox',
                'docker', 'hyperv']
keep = True

def is_json(json_string):
  try:
    json_object = json.loads(json_string)
  except ValueError, e:
    return False
  return True

def decompress_box(box, outputdir=None):

    tar = tarfile.open(box)
    tar.extractall(outputdir)
    tar.close()

def sha1_file(filename):
   #This function returns the SHA-1 hash of (filename)

   checksum = hashlib.sha1()

   with open(filename,'rb') as file:

       chunk = 0
       while chunk != b'':
           chunk = file.read(1024)
           checksum.update(chunk)

   return checksum.hexdigest()

def md5_file(filename):
   #This function returns the MD5 hash of (filename)

   checksum = hashlib.md5()

   with open(filename,'rb') as file:

       chunk = 0
       while chunk != b'':
           chunk = file.read(1024)
           checksum.update(chunk)

   return checksum.hexdigest()

def get_metadata(url):

    try:
        response = urlopen(url)
    except HTTPError as e:
        if e.code == 404:
            print "METADATA NOT FOUND"
            sys.exit(1)
        else:
            print "--->>>> Couldn't connect to server to get current metadata. <<<<---"
            print "Server failed to fulfill request. Error code: %s" % e.code
            sys.exit(1)
    except URLError as e:
        print "--->>>> Couldn't connect to server to get current metadata. <<<<---"
        print "Failed to connect to server. Reason: %s" % e.reason
        sys.exit(1)
    else:
        output = response.read().decode("utf-8")
        if is_json(output):
            metadata = json.loads(output)
        else:
            print "--->>>> URL isn't a valid JSON file, please check URL. <<<<---"
            sys.exit(1)

    return metadata

def get_latestbox(metadata, selected_provider):

    version = "0.0.0"
    i = 0
    for entry in metadata['versions']:
        if version < entry['version']:
            version = entry['version']
            version_id = i
        i = i + 1

    for provider in metadata['versions'][version_id]["providers"]:
        if selected_provider == provider['name']:
            box = provider

    return box

def get_box(url, output_dir=None):

    if output_dir is None:
        output_dir = os.getcwd()

    output_file = os.path.basename(urlparse.urlsplit(url).path)

    try:
        response = urlopen(url)
    except HTTPError as e:
        if e.code == 404:
            print "BOX FILE NOT FOUND!"
            sys.exit(1)
        else:
            print "--->>>> Couldn't connect to server to get box file. <<<<---"
            print "Server failed to fulfill request. Error code: %s" % e.code
            sys.exit(1)
    except URLError as e:
        print "--->>>> Couldn't connect to server to get box file. <<<<---"
        print "Failed to connect to server. Reason: %s" % e.reason
        sys.exit(1)
    else:
        with open(output_dir + '/' + output_file, 'wb') as file:
            block_size = 8192
            while True:
                buffer = response.read(block_size)
                if not buffer:
                    break

                file.write(buffer)

            file.close()

    fullpath = output_dir + '/' + output_file

    return fullpath

#-------------------------------------------------------------------------------
parser = argparse.ArgumentParser(description="This script downloads latest version of a vagrant box.")
parser.add_argument("--url", "-u", required=True, metavar="url", help="URL of the box you want to download.")
parser.add_argument("--provider", "-p", metavar="provider", help="Default provider for the box you want to download (i.e vmware) (Default is virtualbox).")
parser.add_argument("--outputdir", "-o", metavar="outputdir", help="Path to the output folder where you want to store the download box (Default is current directory).")
parser.add_argument("-d", action='store_true', help="Use this flag if you want to decompress the downloaded box (Default no).")
parser.add_argument("-k", action='store_true', help="Use this flag if you want to keep the downloaded box in case you use the --decompress option (Default no).")

args = parser.parse_args()

if len(sys.argv) > 1:
    if url(args.url):
        metadata = get_metadata(args.url)
        if args.provider is not None:
            for provider in vm_providers:
                if provider == args.provider.lower():
                    selected_provider = args.provider.lower()
                    box = get_latestbox(metadata, selected_provider)

                    if args.outputdir is not None:
                        if os.path.isdir(args.outputdir):
                            boxfile = get_box(box['url'], args.outputdir)
                            if box['checksum_type'] == 'sha1':
                                checksum = sha1_file(boxfile)
                            elif box['checksum_type'] == 'md5':
                                checksum = md5_file(boxfile)
                            else:
                                print "Checksum not supported"
                                sys.exit(1)

                            if checksum == box['checksum']:
                                print "Box downloaded and validated"
                                if args.d is not None:
                                    if args.d:
                                        decompress_box(boxfile, args.outputdir)
                                        if args.k is not None:
                                            if args.k:
                                                print "Downloaded box has been kept"
                                            else:
                                                os.remove(boxfile)
                            else:
                                print "Box file checksum not valid"
                                sys.exit(1)
                        else:
                            print "Outputdir isn't a valid path"
                            sys.exit(1)
                    else:
                        boxfile = get_box(box['url'])
                        if box['checksum_type'] == 'sha1':
                            checksum = sha1_file(boxfile)
                        elif box['checksum_type'] == 'md5':
                            checksum = md5_file(boxfile)
                        else:
                            print "Checksum not supported"
                            sys.exit(1)

                        if checksum == box['checksum']:
                            print "Box downloaded and validated"
                            if args.d is not None:
                                if args.d:
                                    decompress_box(boxfile)
                                    if args.k is not None:
                                        if args.k:
                                            print "Downloaded box has been kept"
                                        else:
                                            os.remove(boxfile)
                        else:
                            print "Box file checksum not valid"
                            sys.exit(1)

        else:
            selected_provider = default_provider
            box = get_latestbox(metadata, selected_provider)

            boxfile = get_box(box['url'])
            if box['checksum_type'] == 'sha1':
                checksum = sha1_file(boxfile)
            elif box['checksum_type'] == 'md5':
                checksum = md5_file(boxfile)
            else:
                print "Checksum not supported"
                sys.exit(1)

            if checksum == box['checksum']:
                if args.d is not None:
                    if args.d:
                        decompress_box(boxfile, args.outputdir)
                        if args.k is not None:
                            if args.k:
                                print "Downloaded box has been kept"
                            else:
                                os.remove(boxfile)
    else:
        print "Entered URL is not valid, please check and try again"
        sys.exit(1)


vagrant-uploader.py
#!/usr/bin/env python

import os, sys, argparse, time, tempfile, re, json, shutil
from urllib2 import Request, urlopen, URLError, HTTPError
from hashlib import sha1
from collections import OrderedDict
from pprint import pprint
from fabric.api import run, env, put, hosts, sudo, settings, hide, get, local
from fabric.tasks import execute

#-------------------------------------------------------------------------------
#-- Script settings

env.use_ssh_config = True
default_settings = "config.json"

#-------------------------------------------------------------------------------

class colors:
    HEADER = '\033[95m'
    OKBLUE = '\033[94m'
    OKGREEN = '\033[92m'
    WARNING = '\033[93m'
    FAIL = '\033[91m'
    ENDC = '\033[0m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'

def display_error(error):
    print colors.FAIL + error + colors.ENDC


def box_name_parser(name):
    result = {}
    boxname_regex = "(?P<company>(?:^[a-zA-Z0-9\-\_\.]*))\/(?P<name>(?:[a-zA-Z0-9\-\_\.]*))"

    regex = re.compile(boxname_regex)
    if regex.match(name) is not None:
        regex = re.match(boxname_regex, name)
        result["company"] = regex.group('company')
        result["name"] = regex.group('name')

    return result


def read_defaults(settings="metadata.json"):
    if os.path.isfile(settings):
        with open(settings) as defaults_file:
            defaults = json.load(defaults_file)
            provider = defaults.get("provider")
            file = defaults.get("file")
            baseurl = defaults.get("baseurl")
            if provider is None:
                display_error("Provider not defined in metadata file, this is a required field.")
                display_error("--->>>> Please set provider value in metadata file <<<<---")
                sys.exit(1)
            if file is None:
                display_error("File not defined in metadata file, this is a required field.")
                display_error("--->>>> Please set file value in metadata file <<<<---")
                sys.exit(1)
            if baseurl is None:
                display_error("Base URL not defined in metadata file, this is a required field.")
                display_error("--->>>> Please set baseurl value in metadata file <<<<---")
                sys.exit(1)

            if not file.lower().endswith('.box'):
                display_error("Box filename needs to ends with .box")
                display_error("--->>>> Please change box filename in metadata file <<<<---")
                sys.exit(1)

            return defaults


def generate_metadata(box, metadata=None):

    if metadata is not None:
        version = {}
        version["version"] = box["version"]

        entry = {}
        entry["name"] = box["provider"]
        entry["url"] = box["baseurl"] + "/" + box["company"] + "/" + box["name"] + "/boxes/" + box["file"]
        entry["checksum_type"] = "sha1"
        entry["checksum"] = sha1(box["file"]).hexdigest()

        providers = version
        providers["providers"] = [entry]

        metadata["versions"] = metadata["versions"] + [providers]

    else:
        metadata = {}
        metadata["name"] = box["company"] + "/" + box["name"]
        metadata["description"] = box["description"]

        version = {}
        version["version"] = box["version"]

        entry = {}
        entry["name"] = box["provider"]
        entry["url"] = box["baseurl"] + "/" + box["company"] + "/" + box["name"] + "/boxes/" + box["file"]
        entry["checksum_type"] = "sha1"
        entry["checksum"] = sha1(box["file"]).hexdigest()

        providers = version
        providers["providers"] = [entry]

        metadata["versions"] = [providers]

    return metadata

#-------------------------------------------------------------------------------

#-- Core script
parser = argparse.ArgumentParser(description="This script uploads the .box file to server on specified path and update metadata file.")
parser.add_argument("--config", "-c", metavar="config", help="Full path to defaults settings config file (In json format, check config.json)")
parser.add_argument("--name", "-n", metavar="name", help="The name of the box in the following syntax company/boxname")
parser.add_argument("--file", "-f", metavar="file", help="The path to the .box file i.e /home/user/mybox.box")
parser.add_argument("--provider", "-p", metavar="provider", help="The name of the provider for the box.")
parser.add_argument("--description", "-d", metavar="description", help="The description of the box.")
parser.add_argument("--version", "-v", metavar="version", help="The version number of the box.")
parser.add_argument("--baseurl", "-b", metavar="baseurl", help="The base URL where box is going to be served.")
parser.add_argument("--serverpath", "-s", metavar="serverpath", help="Full path to server in the following format username@fqdn.domain.com:/path/to/webroot")

args = parser.parse_args()

# We check if --configfile is configured and use that instead of default name
if args.config is not None:
    if os.path.isfile(args.config):
        default_settings = args.config
    else:
        display_error("Configuration file is not a file, please check config file path")
        sys.exit(1)

# We use default settings file if present
if os.path.isfile(default_settings):
    if read_defaults(default_settings) is not None:
        box = read_defaults(default_settings)

        # Parse boxname in format company/boxname and add it to box dict
        if box.get("name") is not None:
            parsedname = box_name_parser(box.get("name"))
            box["company"] = parsedname.get("company")
            box["name"] = parsedname.get("name")
        else:
            box["company"] = "undefined"
            box["name"] = "undefined"

        # We create a temp folder where we are going to copy box and metadata
        # before uploading it to server
        if os.path.isfile(box.get("file")):
            # Create a randomic temporary folder
            tempfolder = tempfile.mkdtemp()
            directory = os.path.realpath(os.path.join(tempfolder, box.get("company"), box.get("name")))
            os.makedirs(directory)

            # Copy box to temp folder
            boxes = os.path.realpath(os.path.join(directory, "boxes"))
            os.makedirs(boxes)
            shutil.copy(box.get("file"), boxes)

            metadata_url = box.get("baseurl") + "/" + box.get("company") + "/" + box.get("name")
            remote_metadata = Request(metadata_url)
            try:
                response = urlopen(remote_metadata)
            except HTTPError as e:
                if e.code == 404:
                    # This is first upload of box
                    metadata = generate_metadata(box)

                    with open(directory + '/metadata.json', 'w') as output_file:
                        json.dump(metadata, output_file, sort_keys=True, indent=2, separators=(',', ': '))

                    with hide('output','running','warnings'), settings(warn_only=True, host_string=box["server"]):
                        result = put(tempfolder + "/" + box.get("company"), box.get("remotepath"))

                        if result.failed:
                            display_error("--->>>> UPLOADING BOX FAILED! <<<<---")
                            display_error("Check Internet connection or access to remote server")
                            sys.exit(1)

                    # Delete tempfolder
                    shutil.rmtree(tempfolder)
                else:
                    display_error("--->>>> Couldn't connect to server to get current metadata. <<<<---")
                    display_error("Server failed to fulfill request. Error code: %s" % e.code)
                    sys.exit(1)
            except URLError as e:
                display_error("--->>>> Couldn't connect to server to get current metadata. <<<<---")
                display_error("Failed to connect to server. Reason: %s" % e.reason)
                sys.exit(1)
            else:
                # Box was uploaded before, we need to update metadata
                # and upload new version if it's not there
                current_metadata = json.load(response)

                # We check if the box version isn't already uploaded/defined
                # in server metadata.json
                already_uploaded = False
                for entry in current_metadata['versions']:
                    if box["version"] == entry['version']:
                        already_uploaded = True

                if not already_uploaded:
                    metadata = generate_metadata(box, current_metadata)

                    with open(directory + '/metadata.json', 'w') as output_file:
                        json.dump(metadata, output_file, sort_keys=True, indent=2, separators=(',', ': '))

                    with hide('output','running','warnings'), settings(warn_only=True, host_string=box["server"]):
                        result = put(tempfolder + "/" + box.get("company"), box.get("remotepath"))

                        if result.failed:
                            display_error("--->>>> UPLOADING BOX FAILED! <<<<---")
                            display_error("Check Internet connection or access to remote server")
                            sys.exit(1)
                else:
                    display_error("--->>>> THIS BOX VERSION WAS ALREADY UPLOADED! <<<<---")
                    display_error("Please check box version")
                    sys.exit(1)

                # Delete tempfolder
                shutil.rmtree(tempfolder)
        else:
            display_error("File path defined in metadata file is not a file.")
            display_error("--->>>> Please correct file value in metadata file <<<<---")
            sys.exit(1)
# If no default settings file is present, use args
else:
    if len(sys.argv) > 1:
        print "Hello there!"
        # TODO
        # Parse args and do same as before
        #else:
        # No args fail hard!


vagrant-upload.py
#!/usr/bin/env python

import os, sys, argparse, time, tempfile, re, json, shutil
from urllib2 import Request, urlopen, URLError, HTTPError
import hashlib
from collections import OrderedDict
from pprint import pprint
from fabric.api import run, env, put, hosts, sudo, settings, hide, get, local
from fabric.tasks import execute

#-------------------------------------------------------------------------------
#-- Script settings

env.use_ssh_config = True
default_settings = "config.json"

#-------------------------------------------------------------------------------

class colors:
    HEADER = '\033[95m'
    OKBLUE = '\033[94m'
    OKGREEN = '\033[92m'
    WARNING = '\033[93m'
    FAIL = '\033[91m'
    ENDC = '\033[0m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'

def display_error(message):
    print colors.FAIL + colors.BOLD + colors.UNDERLINE + message + colors.ENDC


def display_ok(message):
    print colors.OKGREEN + message + colors.ENDC


def boxname_parser(name):
    result = {}
    boxname_regex = "(?P<company>(?:^[a-zA-Z0-9\-\_\.]*))\/(?P<name>(?:[a-zA-Z0-9\-\_\.]*))"

    regex = re.compile(boxname_regex)
    if regex.match(name) is not None:
        regex = re.match(boxname_regex, name)
        result["company"] = regex.group('company')
        result["name"] = regex.group('name')

    return result


def load_settings(settings="metadata.json"):
    if os.path.isfile(settings):
        with open(settings) as defaults_file:
            defaults = json.load(defaults_file)

        return defaults


def sha1_file(filename):
   #This function returns the SHA-1 hash of (filename)

   checksum = hashlib.sha1()

   with open(filename,'rb') as file:

       chunk = 0
       while chunk != b'':
           chunk = file.read(1024)
           checksum.update(chunk)

   return checksum.hexdigest()


def validate_settings(config):

    if config.get("provider") is None:
        display_error("Provider not defined in metadata file, this is a required field.")
        display_error("--->>>> Please set provider value in metadata file <<<<---")
        sys.exit(1)

    if config.get("file") is None:
        display_error("File not defined in metadata file, this is a required field.")
        display_error("--->>>> Please set file value in metadata file <<<<---")
        sys.exit(1)
    else:
        if os.path.isfile(config.get("file")):
            if not config.get("file").lower().endswith('.box'):
                display_error("Box filename needs to ends with .box")
                display_error("--->>>> Please change box filename in metadata file <<<<---")
                sys.exit(1)
        else:
                display_error("Box filename path is not a file!")
                display_error("--->>>> Please change box filename path in metadata file <<<<---")
                sys.exit(1)

    if config.get("baseurl") is None:
        display_error("Base URL not defined in metadata file, this is a required field.")
        display_error("--->>>> Please set baseurl value in metadata file <<<<---")
        sys.exit(1)

    if config.get("remotepath") is None:
        display_error("Remote path not defined in metadata file, this is a required field.")
        display_error("--->>>> Please set remote path value in metadata file <<<<---")
        sys.exit(1)

    if config.get("server") is None:
        display_error("Server not defined in metadata file, this is a required field.")
        display_error("--->>>> Please set server value in metadata file <<<<---")
        sys.exit(1)

    if config.get("description") is None:
        config["description"] = "No description provided."

    if config.get("name") is not None:
        parsedname = boxname_parser(config.get("name"))
        config["company"] = parsedname.get("company")
        config["name"] = parsedname.get("name")
    else:
        config["company"] = "undefined"
        config["name"] = "undefined"

    return config

def generate_metadata(box, metadata=None):

    if metadata is not None:
        version = {}
        version["version"] = box["version"]

        entry = {}
        entry["name"] = box["provider"]
        entry["url"] = box["baseurl"] + "/" + box["company"] + "/" + box["name"] + "/boxes/" + os.path.basename(box["file"])
        entry["checksum_type"] = "sha1"
        entry["checksum"] = sha1_file(box["file"])

        providers = version
        providers["providers"] = [entry]

        metadata["versions"] = metadata["versions"] + [providers]

    else:
        metadata = {}
        metadata["name"] = box["company"] + "/" + box["name"]
        metadata["description"] = box["description"]

        version = {}
        version["version"] = box["version"]

        entry = {}
        entry["name"] = box["provider"]
        entry["url"] = box["baseurl"] + "/" + box["company"] + "/" + box["name"] + "/boxes/" + os.path.basename(box["file"])
        entry["checksum_type"] = "sha1"
        entry["checksum"] = sha1_file(box["file"])

        providers = version
        providers["providers"] = [entry]

        metadata["versions"] = [providers]

    return metadata

#-------------------------------------------------------------------------------

#-- Core script
parser = argparse.ArgumentParser(description="This script uploads the .box file to server on specified path and update metadata file.")
parser.add_argument("--config", "-c", metavar="config", help="Full path to defaults settings config file (In json format, check config.json)")
parser.add_argument("--name", "-n", metavar="name", help="The name of the box in the following syntax company/boxname")
parser.add_argument("--file", "-f", metavar="file", help="The path to the .box file i.e /home/user/mybox.box")
parser.add_argument("--provider", "-p", metavar="provider", help="The name of the provider for the box.")
parser.add_argument("--description", "-d", metavar="description", help="The description of the box.")
parser.add_argument("--version", "-v", metavar="version", help="The version number of the box.")
parser.add_argument("--baseurl", "-b", metavar="baseurl", help="The base URL where box is going to be served.")
parser.add_argument("--serverpath", "-s", metavar="serverpath", help="Full path to server in the following format username@fqdn.domain.com:/path/to/webroot")

args = parser.parse_args()

# We check if --configfile is configured and use that instead of default name
if args.config is not None:
    if os.path.isfile(args.config):
        default_settings = args.config
    else:
        display_error("Configuration file is not a file, please check config file path")
        sys.exit(1)

# We use default settings file if present
if os.path.isfile(default_settings):
    if load_settings(default_settings) is not None:
        # Load settings
        display_ok("Loading settings...")
        config = load_settings(default_settings)

        box = validate_settings(config)

        # We create a temp folder where we are going to copy box and metadata
        # before uploading it to server
        if os.path.isfile(box.get("file")):
            display_ok("Creating temp directory...")
            # Create a randomic temporary folder
            tempfolder = tempfile.mkdtemp()
            directory = os.path.realpath(os.path.join(tempfolder, box.get("company"), box.get("name")))
            os.makedirs(directory)

            display_ok("Copying box file...")
            # Copy box to temp folder
            boxes = os.path.realpath(os.path.join(directory, "boxes"))
            os.makedirs(boxes)
            shutil.copy(box.get("file"), boxes)

            display_ok("Checking current metadata...")
            metadata_url = box.get("baseurl") + "/" + box.get("company") + "/" + box.get("name")
            remote_metadata = Request(metadata_url)
            try:
                response = urlopen(remote_metadata)
            except HTTPError as e:
                if e.code == 404:
                    display_ok("No current metadata found, creating it...")
                    # This is first upload of box
                    metadata = generate_metadata(box)

                    with open(directory + '/metadata.json', 'w') as output_file:
                        json.dump(metadata, output_file, sort_keys=True, indent=2, separators=(',', ': '))

                    display_ok("Uploading box and metadata to repository...")
                    with hide('output','running','warnings'), settings(warn_only=True, host_string=box["server"]):
                        result = put(tempfolder + "/" + box.get("company"), box.get("remotepath"))

                        if result.failed:
                            display_error("--->>>> UPLOADING BOX FAILED! <<<<---")
                            display_error("Check Internet connection or access to remote server")
                            sys.exit(1)
                        else:
                            display_ok("Box " + box.get("company") + "/" + box.get("name") + " uploaded successfully!")

                    # Delete tempfolder
                    shutil.rmtree(tempfolder)
                else:
                    display_error("--->>>> Couldn't connect to server to get current metadata. <<<<---")
                    display_error("Server failed to fulfill request. Error code: %s" % e.code)
                    sys.exit(1)
            except URLError as e:
                display_error("--->>>> Couldn't connect to server to get current metadata. <<<<---")
                display_error("Failed to connect to server. Reason: %s" % e.reason)
                sys.exit(1)
            else:
                display_ok("Updating box " + box.get("company") + "/" + box.get("name") + "...")
                # Box was uploaded before, we need to update metadata
                # and upload new version if it's not there
                current_metadata = json.load(response)

                # We check if the box version isn't already uploaded/defined
                # in server metadata.json
                already_uploaded = False
                for entry in current_metadata['versions']:
                    if box["version"] == entry['version']:
                        already_uploaded = True

                if not already_uploaded:
                    metadata = generate_metadata(box, current_metadata)

                    with open(directory + '/metadata.json', 'w') as output_file:
                        json.dump(metadata, output_file, sort_keys=True, indent=2, separators=(',', ': '))

                    display_ok("Uploading box and metadata to repository...")
                    with hide('output','running','warnings'), settings(warn_only=True, host_string=box["server"]):
                        result = put(tempfolder + "/" + box.get("company"), box.get("remotepath"))

                        if result.failed:
                            display_error("--->>>> UPLOADING BOX FAILED! <<<<---")
                            display_error("Check Internet connection or access to remote server")
                            sys.exit(1)
                        else:
                            display_ok("Box " + box.get("company") + "/" + box.get("name") + " uploaded successfully!")

                else:
                    display_error("--->>>> THIS BOX VERSION WAS ALREADY UPLOADED! <<<<---")
                    display_error("Please check box version")
                    sys.exit(1)

                # Delete tempfolder
                shutil.rmtree(tempfolder)
        else:
            display_error("File path defined in metadata file is not a file.")
            display_error("--->>>> Please correct file value in metadata file <<<<---")
            sys.exit(1)
# If no default settings file is present, use args
else:
    if len(sys.argv) > 1:
        print "Hello there!"
        # TODO
        # Parse args and do same as before
        #else:
        # No args fail hard!
